src/transformStateWithClones.js

```diff
@@ -6,8 +6,33 @@
  *
  * @return {Object[]}
  */
+/**
+ * @param {Object} state
+ * @param {Object[]} actions
+ */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const history = [];
+
+  actions.forEach((action, index) => {
+    let newState = index === 0 ? { ...state } : { ...history[index - 1] };
+
+    switch (action.type) {
+      case 'clear':
+        newState = {};
+        break;
+      case 'addProperties':
+        Object.assign(newState, action.extraData);
+        break;
+      case 'removeProperties':
+        if (Array.isArray(action.keysToRemove)) {
```

Comment:
to avoid errors you should prevent the case where there's no property action.keysToRemove and also action.extraData